src/mesh/refinement_cc_in_one.hpp

```diff
@@ -36,6 +37,170 @@ TaskStatus RestrictPhysicalBounds(MeshData<Real> *md);
+  const Real fine001 = fine(l, m, n, k, j, i + 1);
+  Real vol010, vol011, vol100, vol101, vol110, vol111;
+  Real fine010, fine011, fine100, fine101, fine110, fine111;
+  vol010 = vol011 = vol100 = vol101 = vol110 = vol111 = 0;
+  fine010 = fine011 = fine100 = fine101 = fine110 = fine111 = 0;
+  if (DIM > 1) {
+    vol010 = coords.Volume(k, j + 1, i);
+    vol011 = coords.Volume(k, j + 1, i + 1);
+    fine010 = fine(l, m, n, k, j + 1, i);
+    fine011 = fine(l, m, n, k, j + 1, i + 1);
+  }
+  if (DIM > 2) {
+    vol100 = coords.Volume(k + 1, j, i);
+    vol101 = coords.Volume(k + 1, j, i + 1);
+    vol110 = coords.Volume(k + 1, j + 1, i);
+    vol111 = coords.Volume(k + 1, j + 1, i + 1);
+    fine100 = fine(l, m, n, k + 1, j, i);
+    fine101 = fine(l, m, n, k + 1, j, i + 1);
+    fine110 = fine(l, m, n, k + 1, j + 1, i);
+    fine111 = fine(l, m, n, k + 1, j + 1, i + 1);
+  }
+  // KGF: add the off-centered quantities first to preserve FP
+  // symmetry
+  const Real tvol =
+      ((vol000 + vol010) + (vol001 + vol011)) + ((vol100 + vol110) + (vol101 + vol111));
+  coarse(l, m, n, ck, cj, ci) =
+      (((fine000 * vol000 + fine010 * vol010) + (fine001 * vol001 + fine011 * vol011)) +
+       ((fine100 * vol100 + fine110 * vol110) + (fine101 * vol101 + fine111 * vol111))) /
+      tvol;
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+Real GradMinMod(const Real ccval, const Real fm, const Real fp, const Real dxm,
+                const Real dxp) {
+  const Real gxm = (ccval - fm) / dxm;
+  const Real gxp = (fp - ccval) / dxp;
+  return 0.5 * (SIGN(gxm) + SIGN(gxp)) * std::min(std::abs(gxm), std::abs(gxp));
+}
+
+// TODO(JMM): this could be simplified if grid spacing was always uniform
+template <int DIM>
+KOKKOS_FORCEINLINE_FUNCTION Real GetX(const Coordinates_t &coords, int i) {
+  if (DIM == 1) return coords.x1v(i);
+  if (DIM == 2) return coords.x2v(i);
+  if (DIM == 3) return coords.x3v(i);
+  PARTHENON_FAIL("Unknown dimension");
+}
+template <int DIM>
+KOKKOS_FORCEINLINE_FUNCTION void
+GetSlopes(const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+          const IndexRange &cib, const IndexRange &ib, int i, int &fi, Real &dxm,
+          Real &dxp, Real &dxfm, Real &dxfp) {
+  fi = (i - cib.s) * 2 + ib.s;
+  const Real xm = GetX<DIM>(coarse_coords, i - 1);
+  const Real xc = GetX<DIM>(coarse_coords, i);
+  const Real xp = GetX<DIM>(coarse_coords, i + 1);
+  dxm = xc - xm;
+  dxp = xp - xc;
+  const Real fxm = GetX<DIM>(coords, fi);
+  const Real fxp = GetX<DIM>(coords, fi + 1);
+  dxfm = xc - fxm;
+  dxfp = fxp - xc;
+}
+
+// TODO(JMM): This function signature is real gross. Might be worth
+// coalescing some of this stuff into structs, rather than unpacking
+// it and then passing it in.
+
+// TODO(JMM): Compared to the previous version of the code, this one
+// multiplies by zero sometimes.
+
+// In the absence of constexpr if, use template DIM to resolve ifs at
+// compile time and avoid branching.
```

Comment:
Not sure I understand the comment. Is this a todo? I still see `if (DIM > 1)` etc in the function that could eventually be  `constexpr`

Reply:
Yeah this is essentially the TODO, make the ifs constexpr in the future.

Reply:
But you can check that if statements on template parameters do get compiled out.